### DISCUSSION TITLE
canbus remote request

### DISCUSSION THREAD

**Discussion by eiten** [zephyrproject-rtos/zephyr#78994] [Q&A] [answered]
Hello everybody,

how is it possible to 
1. Send rtr with the current zephyr version? can_frame has not rtr-field anymore
2. Receive rtr and then send the requested data? If you set the rx filter to an ID, rtr will not trigger the callback/message queue.

Thanks, Edi

**Comment by henrikbrixandersen** [comment, marked as answer]
>     1. Send rtr with the current zephyr version? can_frame has not rtr-field anymore

You set the `CAN_FRAME_RTR` flag in `flags` member in the `struct can_frame` TX instance.
 
>     2. Receive rtr and then send the requested data? If you set the rx filter to an ID, rtr will not trigger the callback/message queue.

You need to enable `CONFIG_CAN_ACCEPT_RTR=y` to allow incoming RTR frames, add a CAN RX filter for the specific CAN ID and check for the `CAN_FRAME_RTR` flag in matching RX frames in your callback. If the CAN ID matches and the RTR flag is present, you need to queue up a response TX frame (which can be prepared in advance, if needed).

Hardware support for RTR is very, very different between various CAN controllers (quite a few controllers don't even support RTR and its use is discouraged), making it very difficult to come up with a common Zephyr CAN API abstraction for handling automatic RTR replies.

**Reply by eiten** [reply]
Hi @henrikbrixandersen 

Thank you very much, I was missing the CONFIG_CAN_ACCEPT_RTR=y part! Now everything works as expected.!

Have a nice day!

**Reply by henrikbrixandersen** [reply]
> Hi @henrikbrixandersen 
> 
> 
> 
> Thank you very much, I was missing the CONFIG_CAN_ACCEPT_RTR=y part! Now everything works as expected.!

Glad you got it to work. 

> Have a nice day!

Thanks, same to you ☺️
